### PR TITLE
[Arm64] Use ldp/stp in genCodeForCpObj()

### PR DESCRIPTION
**Pull request by sdmaclea**

### CODE REVIEW

**Comment by sdmaclea** [comment]
@dotnet/jit-contrib @dotnet/arm64-contrib PTAL Passes pr1 tests on Arm64/Unix

File: src/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -3461,19 +3461,21 @@ void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
...
 
+    unsigned slots = cpObjNode->gtSlots;
+
     // Temp register used to perform the sequence of loads and stores.
-    regNumber tmpReg = cpObjNode->GetSingleTempReg();
-    assert(genIsValidIntReg(tmpReg));
+    regNumber tmpReg = cpObjNode->ExtractTempReg();
+    regNumber tmpReg2 = (slots > 1) ? cpObjNode->GetSingleTempReg() : REG_NA;
+    assert(genIsValidIntReg(tmpReg) && (tmpReg != REG_WRITE_BARRIER_SRC_BYREF) && (tmpReg != REG_WRITE_BARRIER_DST_BYREF));
+    assert((slots == 1) || (genIsValidIntReg(tmpReg2) && (tmpReg2 != tmpReg) && (tmpReg2 != REG_WRITE_BARRIER_SRC_BYREF) && (tmpReg2 != REG_WRITE_BARRIER_DST_BYREF)));
```
Comment (briansull) [inline src/jit/codegenarm64.cpp]: These compound asserts are not the best; as when they fire you don't immediately know what the problem is.
I would rewrite this as:

```
if (slots == 1)
{
   // stuff to do when slots are 1 and the asserts that apply 
}
else
{
   // stuff to do when slots are 2 or greater and the asserts that apply 
}
```

File: src/jit/codegenarm64.cpp [hunk 2]
```diff
@@ -3482,10 +3484,21 @@ void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
             else if (gcPtrs[i] == GCT_BYREF)
                 attr = EA_BYREF;
 
-            emit->emitIns_R_R_I(INS_ldr, attr, tmpReg, REG_WRITE_BARRIER_SRC_BYREF, TARGET_POINTER_SIZE,
-                                INS_OPTS_POST_INDEX);
-            emit->emitIns_R_R_I(INS_str, attr, tmpReg, REG_WRITE_BARRIER_DST_BYREF, TARGET_POINTER_SIZE,
-                                INS_OPTS_POST_INDEX);
+            if((i + 1 < slots) && (gcPtrs[i] == gcPtrs[i + 1]))
```
Comment (briansull) [inline src/jit/codegenarm64.cpp]: Add a comment here:
`// Check if we have two or more remaining slots and use a ldp/stp sequence`

File: src/jit/codegenarm64.cpp [hunk 3]
```diff
@@ -3482,10 +3484,21 @@ void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
...
+            {
+                emit->emitIns_R_R_R_I(INS_ldp, attr, tmpReg, tmpReg2, REG_WRITE_BARRIER_SRC_BYREF, 2*TARGET_POINTER_SIZE,
+                                      INS_OPTS_POST_INDEX);
+                emit->emitIns_R_R_R_I(INS_stp, attr, tmpReg, tmpReg2, REG_WRITE_BARRIER_DST_BYREF, 2*TARGET_POINTER_SIZE,
+                                      INS_OPTS_POST_INDEX);
+                ++i;
+            }
+            else
```
Comment (briansull) [inline src/jit/codegenarm64.cpp]: Add a comment
`else  // copy the last remaining (or only) value using ldr/str `

File: src/jit/codegenarm64.cpp [hunk 4]
```diff
@@ -3482,10 +3484,21 @@ void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
...
+            {
+                emit->emitIns_R_R_I(INS_ldr, attr, tmpReg, REG_WRITE_BARRIER_SRC_BYREF, TARGET_POINTER_SIZE,
+                                    INS_OPTS_POST_INDEX);
+                emit->emitIns_R_R_I(INS_str, attr, tmpReg, REG_WRITE_BARRIER_DST_BYREF, TARGET_POINTER_SIZE,
+                                    INS_OPTS_POST_INDEX);
+            }
```
Comment (briansull) [inline src/jit/codegenarm64.cpp]: You could even add a `break `here as well, if you want..

If you do that it probably makes more sense to write the loop as a 
```
do {
} while (i < slots)
```

File: src/jit/codegenarm64.cpp [hunk 5]
```diff
@@ -3498,11 +3511,21 @@ void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
             switch (gcPtrs[i])
             {
                 case TYPE_GC_NONE:
-                    // TODO-ARM64-CQ: Consider using LDP/STP to save codesize in case of contigous NON-GC slots.
-                    emit->emitIns_R_R_I(INS_ldr, EA_8BYTE, tmpReg, REG_WRITE_BARRIER_SRC_BYREF, TARGET_POINTER_SIZE,
-                                        INS_OPTS_POST_INDEX);
-                    emit->emitIns_R_R_I(INS_str, EA_8BYTE, tmpReg, REG_WRITE_BARRIER_DST_BYREF, TARGET_POINTER_SIZE,
-                                        INS_OPTS_POST_INDEX);
+                    if((i + 1 < slots) && (gcPtrs[i] == gcPtrs[i + 1]))
```
Comment (briansull) [inline src/jit/codegenarm64.cpp]: Add a comment:
`// Check if the next slot's type is also TYP_GC_NONE and use ldp/stp`

File: src/jit/codegenarm64.cpp [hunk 6]
```diff
@@ -3498,11 +3511,21 @@ void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
...
+                    {
+                        emit->emitIns_R_R_R_I(INS_ldp, EA_8BYTE, tmpReg, tmpReg2, REG_WRITE_BARRIER_SRC_BYREF, 2*TARGET_POINTER_SIZE,
+                                              INS_OPTS_POST_INDEX);
+                        emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, tmpReg, tmpReg2, REG_WRITE_BARRIER_DST_BYREF, 2*TARGET_POINTER_SIZE,
+                                              INS_OPTS_POST_INDEX);
+                        ++i;
+                    }
+                    else
+                    {
```
Comment (briansull) [inline src/jit/codegenarm64.cpp]: Add ToDo comment:
`// ToDo: Current limitations only allows using ldp/stp when both of the GC types are  TYP_GC_NONE`

File: src/jit/codegenarm64.cpp [hunk 7]
```diff
@@ -3498,11 +3511,21 @@ void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
...
-                                        INS_OPTS_POST_INDEX);
-                    emit->emitIns_R_R_I(INS_str, EA_8BYTE, tmpReg, REG_WRITE_BARRIER_DST_BYREF, TARGET_POINTER_SIZE,
-                                        INS_OPTS_POST_INDEX);
+                    if((i + 1 < slots) && (gcPtrs[i] == gcPtrs[i + 1]))
+                    {
+                        emit->emitIns_R_R_R_I(INS_ldp, EA_8BYTE, tmpReg, tmpReg2, REG_WRITE_BARRIER_SRC_BYREF, 2*TARGET_POINTER_SIZE,
+                                              INS_OPTS_POST_INDEX);
+                        emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, tmpReg, tmpReg2, REG_WRITE_BARRIER_DST_BYREF, 2*TARGET_POINTER_SIZE,
+                                              INS_OPTS_POST_INDEX);
+                        ++i;
```
Comment (briansull) [inline src/jit/codegenarm64.cpp]: `   ++i;  // extra increment of i, since we are copying two items`

File: src/jit/lsraarmarch.cpp [hunk 8]
```diff
@@ -792,6 +792,14 @@ void Lowering::TreeNodeInfoInitBlockStore(GenTreeBlk* blkNode)
             // a temporary register to perform the sequence of loads and stores.
             blkNode->gtLsraInfo.internalIntCount = 1;
 
+            if (size >= 2 * REGSIZE_BYTES)
+            {
+                // Use ldp/stp to reduce code size and improve performance
```
Comment (briansull) [inline src/jit/lsraarmarch.cpp]:
```
// We will use ldp/stp to reduce code size and improve performance
// so we need to reserve an extra internal register
```

File: src/jit/lsraarmarch.cpp [hunk 9]
```diff
@@ -792,6 +792,14 @@ void Lowering::TreeNodeInfoInitBlockStore(GenTreeBlk* blkNode)
...
+                blkNode->gtLsraInfo.internalIntCount++;
+            }
+
+            blkNode->gtLsraInfo.setInternalCandidates(l, RBM_ALLINT & ~(RBM_WRITE_BARRIER_DST_BYREF | RBM_WRITE_BARRIER_SRC_BYREF));
```
Comment (briansull) [inline src/jit/lsraarmarch.cpp]: `// We can't use the special Write Barrier registers, so exclude them from the mask`

File: src/jit/lsraarmarch.cpp [hunk 10]
```diff
@@ -792,6 +792,14 @@ void Lowering::TreeNodeInfoInitBlockStore(GenTreeBlk* blkNode)
...
+
             dstAddr->gtLsraInfo.setSrcCandidates(l, RBM_WRITE_BARRIER_DST_BYREF);
             // If we have a source address we want it in REG_WRITE_BARRIER_SRC_BYREF.
```
Comment (briansull) [inline src/jit/lsraarmarch.cpp]: `If we have a source address we want it to be in REG_WRITE_BARRIER_SRC_BYREF.`

File: src/jit/codegenarm64.cpp (hunk 6 above)
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Is that correct?

> ```// ToDo: Current limitations only allows using ldp/stp when both of the GC types are TYP_GC_NONE```
I thought it was 
```
// ToDo: Current limitations only allows using ldp/stp when both of the GC types match
```

Comment (briansull) [inline src/jit/codegenarm64.cpp]: Yeah, that is probably the correct comment for the current limitation.
however your implementation only allows for two TYPE_GC_NONE to utilize the ldp/stp.

Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: >  your implementation only allows for two TYPE_GC_NONE to utilize the ldp/stp

That's not true.  
In the cpObj which may not be on stack case.  It only allow ldr/str or ldp/stp if ASSIGN_BYREF is not required.

In the stack case it only allows if the type matches.

**Comment by sdmaclea** [comment]
@briansull This should address your comments.

File: src/jit/lsraarmarch.cpp [hunk 11]
```diff
@@ -792,7 +792,20 @@ void Lowering::TreeNodeInfoInitBlockStore(GenTreeBlk* blkNode)
...
+                // We will use ldp/stp to reduce code size and improve performance
+                // so we need to reserve an extra internal register
+                blkNode->gtLsraInfo.internalIntCount++;
+            }
+
+            // We can't use the special Write Barrier registers, so exclude them from the mask
+            regMaskTP internalIntCandidates = RBM_ALLINT & ~(RBM_WRITE_BARRIER_DST_BYREF | RBM_WRITE_BARRIER_SRC_BYREF);
+            blkNode->gtLsraInfo.setInternalCandidates(l, internalIntCandidates);
+
+            // If we have a source address we want it in RBM_WRITE_BARRIER_DST_BYREF.
```
Comment (briansull) [inline src/jit/lsraarmarch.cpp]: I think that this comment should read:
` // If we have a dest address we want it in RBM_WRITE_BARRIER_DST_BYREF.`

**Pull request merged**